lib/rubocop/cop/rspec/overwriting_setup.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks if there is a let/subject that overwrites an existing one.
+      #
+      # @example
+      #   # bad
+      #     let(:foo) { bar }
+      #     let(:foo) { baz }
+      #
+      #     subject(:foo) { bar }
+      #     let(:foo) { baz }
+      #
+      #     let(:foo) { bar }
+      #     let!(:foo) { baz }
+      #
+      #   # good
+      #     subject(:test) { something }
+      #     let(:foo) { bar }
+      #     let(:baz) { baz }
+      #     let!(:other) { other }
+      class OverwritingSetup < Cop
+        MSG = '`%{name}` is already defined.'.freeze
+
+        def_node_matcher :setup?, <<-PATTERN
+          (block (send nil {:let :let! :subject} (sym $_)) ...)
+        PATTERN
+
+        def on_block(node)
+          return unless example_group?(node)
+
+          _describe, _args, body = *node
+
+          find_duplicates(body) do |duplicate, name|
+            add_offense(duplicate, :expression, format(MSG, name: name))
+          end
+        end
+
+        private
+
+        def find_duplicates(node)
+          already_defined = []
```

Comment:
Could we name this collector after its contents? Perhaps `setup_expressions = []`?